### PR TITLE
Added a lint check for mouse cursor definitions

### LINKED ISSUE #17610: CheckSequenceSprites does not check cursor sequences
```
Exception of type `System.InvalidOperationException`: Cursor does not have a sequence `nuke`
  at OpenRA.Graphics.CursorProvider.GetCursorSequence (System.String cursor) [0x00010] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Graphics/CursorProvider.cs:67 
  at OpenRA.Graphics.SoftwareCursor.Render (OpenRA.Renderer renderer) [0x0002a] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Graphics/SoftwareCursor.cs:107 
  at OpenRA.Game.RenderTick () [0x0017b] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Game.cs:711 
  at OpenRA.Game.Loop () [0x0015a] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Game.cs:833 
  at OpenRA.Game.Run () [0x0003c] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Game.cs:852 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Game.cs:263 
  at OpenRA.Program.Main (System.String[] args) [0x00044] in /home/matthias/Entwicklung/OpenHV/engine/OpenRA.Game/Support/Program.cs:37
```

was not catched.

### PR DESCRIPTION
**Pull request by Mailaender**
Followup of https://github.com/OpenRA/ra2/pull/631 as those errors create crashes that can be hard to spot.
Closes https://github.com/OpenRA/OpenRA/issues/17610.

### CODE REVIEW

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/Minelayer.cs]: Missing `[CursorReference]`?

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/AttackMove.cs]: These are the same as the checks above, so will never run.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/AttackMove.cs]: This doesn't look correct.

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (Mailaender) [reply OpenRA.Mods.Cnc/Traits/Minelayer.cs]: Yes

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/AttackMove.cs]: That was sloppy.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/AttackMove.cs]: I don't have a solution for it and it doesn't seem to regress in-game.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/AttackMove.cs]: The (attack)move-blocked cursor will no longer show when hovering outside of the map bounds.

Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/AttackMove.cs]: I reworked this.

File: OpenRA.Mods.Common/Traits/Cargo.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Cargo.cs:63]: Style nit: these should all go before the `[Desc()]`s.

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/Minelayer.cs]: Please pass `MineLayerInfo` into the `BeginMinefieldOrderTargeter` constructor instead of querying it all the time here.

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/Minelayer.cs]: Please assign a `readonly string cursor` in the constructor like we do for the overlay tiles.

File: OpenRA.Mods.Common/Lint/CheckCursors.cs
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckCursors.cs:33]: FYI: this doesn't account for the custom terrain move cursors defined in tilesets, but this should be fixed in a followup PR (which I may take on after this one has been merged) by changing `MobileInfo.Cursor` to support terrain-specific overrides. Defining them on the tileset isn't great from either a code quality or moddability perspective.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/AttackMove.cs]: This variable no longer holds a prefix, so should be renamed to `isAssaultMove`.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/AttackMove.cs:158]: The logic in this block doesn't look correct. We take the cursor from the first actor, but this actor may not be able to move into the shroud.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/AttackMove.cs]: redundant check.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (pchote) [reply OpenRA.Mods.Common/Traits/AttackMove.cs]: This still doesn't look correct. We shouldn't be introducing "default" at all.

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (Mailaender) [reply OpenRA.Mods.Cnc/Traits/Minelayer.cs]: https://github.com/OpenRA/OpenRA/pull/17843/ already solved this better, so I would like that get merged first.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/AttackMove.cs]: I can't access the `subject.Trait.Info` from below here so this fallback though probably never hitting is required.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/AttackMove.cs]: Return `null` so that the engine has a chance to substitute a cursor from a different unit in the selection.

**Comment by Mailaender** [comment]
Waiting for https://github.com/OpenRA/OpenRA/pull/18110 as well now.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/AttackMove.cs]: Okay.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/AttackMove.cs:158]: Can you explain that further? I just changed the `Any()` for a null check.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/AttackMove.cs:158]: I think the idea here is that the blocked cursor of the actor preventing us from moving into shroud should be used instead of the cursor of the first actor. (This was present before, but not visible since people couldn't set the cursor per unit.)

Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/AttackMove.cs:158]: Sorry I still don't get it. The logic didn't change. I just replaced hardcoded strings/prefixes with variables. `subjects.First()` was checked against before.

**Comment by Mailaender** [comment]
Rebased and proposed a fix for the attack move inconsistency.

**Comment by deleted-user-1** [comment]
Will need in-game testing.

**Comment by pchote** [comment]
It would be good if we could rebase this on #18514 and adjust it to follow the same code patterns.
This provides a template to fix a few issues:
* Use `CursorSequences` instead of manually reparsing yaml files to reduce duplication and maintenance overheads.
* Use `ILintRulesPass` instead of `ILintMapPass` so that the checks apply to the default mod rules, not just maps.
* Add support for cursor references in dictionaries (needed for the followup pr to fix https://github.com/OpenRA/OpenRA/pull/17933#discussion_r414572129).
* Check for missing sprites in `--check-missing-sprites` in addition to the references in `CheckCursors` (which should stay as its own separate lint test).

**Comment by Mailaender** [comment]
Rebased upon your pull request and implemented the first two proposals.

File: OpenRA.Game/Traits/LintAttributes.cs
Comment (pchote) [inline OpenRA.Game/Traits/LintAttributes.cs]: Copy https://github.com/OpenRA/OpenRA/pull/18514/commits/a2f8e8126c8d66e52790e76ee9f6eea252f5cd98#diff-942a43fba9405303fa949f30de960002R43-R51 here.

File: OpenRA.Mods.Common/Lint/CheckCursors.cs
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckCursors.cs]: Change this to 
```csharp
var cursorReference = field.GetCustomAttributes<CursorReferenceAttribute>(true).FirstOrDefault();
if (cursorReference == null)
	continue;
```

and remove the now-unnecessary block level.

File: OpenRA.Mods.Common/Lint/CheckCursors.cs
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckCursors.cs]: Change this to
```csharp
var cursor = LintExts.GetFieldValues(traitInfo, field, emitError, cursorReference.DictionaryReference).FirstOrDefault();
```

**Comment by pchote** [comment]
This hasn't implemented my first point - you are still manually opening and reparsing the yaml files!

**Review by pchote — COMMENTED** [review]
Adding dictionary support (the third point) should be easy:

**Comment by Mailaender** [comment]
I had to revert the use of `CursorProvider` as it crashes Travis.

**Comment by abcdefg30** [comment]
I'm afraid this needs a rebase now.

**Comment by Mailaender** [comment]
Rebased.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (pchote) [reply OpenRA.Mods.Common/Traits/AttackMove.cs:158]: The logic *did* change. The existing code doesn't use `First` for `MoveIntoShroud`, but `Any`.

The effect of this can easily be seen ingame: the current behaviour will always show the blocked cursor variant if you have a mix of ground units and aircraft in TS. The PR behaviour will change depending on whether you built the ground unit or aircraft first (which is what determines `subjects.First`). There is also a crash if you activate attack-move while the cursor is over the shroud.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/AttackMove.cs:158]: The problem is that `subjects` is not a list of actors, but a list of `TraitPair`s. `TraitPair` is a value type, so is never null. The `subject != null` and `shroudAvoider != null` checks are therefore *always* true. The first check can change to `subject.Actor != null` and the second can change to `var cannotMove = subjects.FirstOrDefault(a => !a.Trait.Info.MoveIntoShroud).Trait` (also changing `shroudAvoider`, which is an odd variable name).

Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/AttackMove.cs:158]: I changed things, but I am not sure if I got it right.

**Comment by pchote** [comment]
The latest push is failing the lint tests. Can you please also squash the fixes into the main commit?

**Comment by Mailaender** [comment]
Rebased and fixed the oversight.

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (phrohdoh) [inline OpenRA.Mods.Cnc/Traits/Minelayer.cs]: nit: start vs begin, do we need to use 2 separate terms?

**Review by phrohdoh — COMMENTED** [review]
inline nit (can't seem to comment without creating a review when using the gh mobile client)

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (Mailaender) [reply OpenRA.Mods.Cnc/Traits/Minelayer.cs]: Shortened it.

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/Minelayer.cs]: As far as I can tell, this trait / order generators don't require this actor to have the `Mine` trait?

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/Minelayer.cs]: This is the wrong cursor type!

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/Minelayer.cs]: This is the wrong cursor type!

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/AttackMove.cs]: By process of elimination, `!isAssaultMove` is by always true here, and `!isAssaultMove && blocked` below.

I suggest simplifying this to
```
if (isAssaultMove)
	return blocked ? cannotMove.Info.AssaultMoveBlockedCursor : info.AssaultMoveCursor;

return blocked ? cannotMove.Info.AttackMoveBlockedCursor : info.AttackMoveCursor;
```

**Review by pchote — COMMENTED** [review]
The following traits (and maybe others, but these are the ones I noticed) define cursors that are missing the CursorReference attribute:
* Aircraft
* TransformsIntoAircraft
* TransformsIntoPassenger
* TransformsIntoRepairable
* Carryall
* DeliversCash
* DeliversExperience
* Harvester
* Passenger
* Repairable
* RepairableNear

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (Mailaender) [reply OpenRA.Mods.Cnc/Traits/Minelayer.cs]: You are right. This doesn't fix a crash or anything.

File: OpenRA.Mods.Cnc/Traits/Minelayer.cs
Comment (Mailaender) [reply OpenRA.Mods.Cnc/Traits/Minelayer.cs]: Ah, yes that happened during all the rebases.

File: OpenRA.Mods.Common/Traits/AttackMove.cs
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/AttackMove.cs]: Thanks. That block got way out of hand.

**Comment by Mailaender** [comment]
Added all the missing `"enter"` cursors.

**Review by pchote — DISMISSED** [review]
Changes LGTM now and didn't notice any new issues ingame.

File: OpenRA.Mods.Common/Orders/BeaconOrderGenerator.cs
Comment (reaperrr) [inline OpenRA.Mods.Common/Orders/BeaconOrderGenerator.cs:33]: I didn't follow the entire discussion, was this left for later intentionally?

File: OpenRA.Mods.Common/Orders/DeployOrderTargeter.cs
Comment (reaperrr) [inline OpenRA.Mods.Common/Orders/DeployOrderTargeter.cs]: Same here.

**Comment by reaperrr** [comment]
LGTM if the answer to my question is 'yes'.

File: OpenRA.Mods.Common/Orders/DeployOrderTargeter.cs
Comment (pchote) [reply OpenRA.Mods.Common/Orders/DeployOrderTargeter.cs]: This one is easily fixable. Remove the `public DeployOrderTargeter(string order, int priority)` ctor and add cursor definitions to `PrimaryBuilding` and `MadTank`.

File: OpenRA.Mods.Common/Orders/BeaconOrderGenerator.cs
Comment (pchote) [reply OpenRA.Mods.Common/Orders/BeaconOrderGenerator.cs:33]: This one is not easily fixable because this PR is not set up to lint UI-defined cursors at all. This should be left to a followup PR.

**Comment by Mailaender** [comment]
`PrimaryBuilding` and `MadTank` now also check their cursor and I found an unnecessary trait lookup.

File: OpenRA.Mods.Cnc/Traits/MadTank.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/MadTank.cs]: Showing the standard deploy cursor after deploying was indeed bogus, so good spotting here.

IMO showing a blocked cursor is the wrong solution though, since these cursors usually indicate that something is generally possible but isn't right now. The MAD tank deployment is a one-way transformation, so I don't think we should be showing the deploy cursor as an option at all (equivalent to disabling the trait, if the trait were rewritten/replaced using modern style).

This can be done using
```csharp
if (!initiated)
	yield return new DeployOrderTargeter("Detonate", 5, () => info.DeployCursor);
```
and then we could drop the `DeployBlockedCursor` definition.

Comment (Mailaender) [reply OpenRA.Mods.Cnc/Traits/MadTank.cs]: Oh yes nice, that looks way more polished. Thanks!

**Comment by Mailaender** [comment]
Fixed Mad Tank showing the deployment cursor when the uninterruptible sequence started.

**Review by pchote — DISMISSED** [review]

**Comment by abcdefg30** [comment]
Unfortunately this needs a rebase now. :(

**Comment by Mailaender** [comment]
Rebased.

**Comment by penev92** [comment]
I believe this has a 👍 from @pchote.
Does it have one from @reaperrr though? Does it need another review?

**Comment by pchote** [comment]
Yes... this is still waiting on a second :+1: AFAIK, and now needs another rebase.

**Comment by Mailaender** [comment]
Rebased.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ba94c7de57f9052d1f3fd8556cf32103e79716ac...568c5854ecc7c1a1a5ea37dbbbbb86ed46fcf241

**Pull request merged**
